### DISCUSSION TITLE
Porting to other hardware

### DISCUSSION THREAD

**Discussion by McMuppet** [pyControl/code#120] [General]
Hi All,
  I've been looking into this project for a friend who is a researcher and is having problems keeping his old med-pc xp boxes up and running. I'm unable to find a pyboard near me at a reasonable price, and even if I could I was wondering how difficult you all think porting this to a raspberry pi pico or esp32 would be? Both should be able to run micropython. I'm not sure how if porting the code to accept these boards would just be swapping snippets relating to GPIO pins, or require something more drastic. 

What are your thoughts on the feasibility or difficulty or porting this over to other hardware like the Pico?

**Comment by ThomasAkam** [comment]
Hi @McMuppet,

I have not tried to run pyControl on anything other than a pyboard v1.1 so am not sure how much work would be required.  The [pyb](https://docs.micropython.org/en/latest/library/pyb.html#module-pyb) module, which is used extensively by the pyControl framework code, is only available on pyboard (and possibly other STM32 based boards).  The [machine](https://docs.micropython.org/en/latest/library/machine.html#module-machine) module, which is available across different ports implements most of the same functionallity, but adapting pyControl to use it would take some work and it is posisble there are some bits of the pyb module for which there is no simple replacement in  machine.

When I last contacted micropython about the pyboard supply issues they said they have a new board in development to replace the pyboard v1.1, which should be a drop in replacement, which they expected to have ready in autumn 2023.  I've just contacted them to ask for an update and will post here when I hear back from them.  My current feeling is that if they are likely to have a widely available pyboard back in stock soon then it is not worthwile adapting pyControl to different hardware, but if the supply chain issues don't get resolved soon it will be necessary.

In terms of the hardware itself, I think you would want to use a board that matches or exceeds the pyboard v1.1 in terms of processor speed, ram and storage.  I would also recommend using a board that has an analog output (DAC), as pyControl uses the pyboard DAC for generating audio stimuli.  The raspberry pi pico does not have a DAC so the ESP32 might be a better option - it has two 8 bit DACs and although the pyboard has 12 bit DACs the current audio module only uses 8 bit waveforms so should work with the ESP32.

If you do decide to give pyControl a try on different hardware let me know how you get on.

**Reply by ThomasAkam** [reply]
I heard back from micropython and they are doing a large production run of pyboard v1.1 which should be available from late Feb/March.

**Reply by McMuppet** [reply]
Thanks for the detailed reply! Was there any particular reason or use cases pycontrol was written for a microcontroller board like the pyBoard and not as a general python program to run on any desktop/laptop? Probably has to do with I/O pins being directly available I'd imagine?

Again, I'm new into this and don't want to go down a long road having not understood to basic needs of the field and end up abandoning the work. Appreciate your thoughtful replies!

**Comment by ThomasAkam** [comment]
The main reasons for using a microcontroller running python rather than running task code directly on the computer were:

- Access to I/O pins

- Low cost

- Easy parallelisation to many individual setups controlled from a single computer

- More deterministic / low latency timing due to not having an operating system running in the background.

**Comment by alustig3** [comment]
pyboard v1.1 seems to be back in stock: https://store.micropython.org/product/PYBv1.1
